### ISSUE TITLE
Streamlined Usage of OpenKCC Through KCCMovementEngine Abstraction

### ISSUE THREAD

**Issue by MTrecozzi** [nicholas-maltbie/OpenKCC#136]
**Goal:** Simplify the direct use of KCCMovementEngine

**Use Case:** I created a test scene and Unity Package that has a simple character that simply attempts to move in a direction on the XZ axis

As as starting point, I'd like to be able to use the KCCMovementEngine API alongside a simple KCCConfig to make it so the AI agent moves in each of the cardinal directions while making use of OpenKCC under the hood

See the picture below for the "Intended Behaviour"

![image](https://user-images.githubusercontent.com/47111522/213306557-473280bf-0ff3-4738-ab11-ec4875948d25.png)

- While using the KCCMovementEngine and an Appropriate KCCCOnfig, I expect the character to walk up the stairs when moving on the X axis and to be blocked by the walls while moving in the Z axis
- I also expect the character to fall off the last step of the stairs and start descending if gravity is enabled at the KCCConfig or KCCMovementEngine level

It would be interesting if I could then swap the KCCMovementEngine for a KCCMoleEngine, enabling the character to run Up the walls like the mole would be able to, without changing any code, but I understand if this would also require a swap in KCCConfig as well

Below shows my current test case, where the AI Agent is instructed to desire a velocity of 1 along the X axis, and my non KCC solution simply moves the transform to the calculated position each frame.

![Open-KCC-API-Request](https://user-images.githubusercontent.com/47111522/213307811-f63f20de-2826-4bc3-be04-86c6335c2d81.gif)

- This naïve implementation simply causes the agent to phase through objects

The sample package that I will provide to you demonstrates how I created this script and includes commented sections that demonstrate how I would like to replace my naïve implementation with OpenKCC's Collision Solving.

Here is a snippet of how the code currently works and it includes an example of how I'd like to replace my naïve implementation with an OpenKCC API Call

![image](https://user-images.githubusercontent.com/47111522/213308176-1d6812d1-fe9d-401b-b4b6-ef4ba33de692.png)

```c#
    void FixedUpdate()
    {
        // 1. Get Desired Velocity from Input Source
        Vector3 newVelocity = testInputDirection * testSpeed;

        // 2. Make Arbitrary Modifications to Velociyt Based on Character State
        // newVelocity = speedBoost ? newVelocity * 2 : newVelocity;

        // 3. Apply Movement to Character based on Calculated Velocity
        ApplyMovement(newVelocity, Time.fixedDeltaTime);
    }

    void ApplyMovement(Vector3 velocity, float deltaTime)
    {
        // 1. Temporary, Non Collision Based Solution
        transform.position += velocity * deltaTime;

        // 2. Ideal Solution Using OpenKCC, Benefitting from Collision and
        // Gravity Settings of the KCCMovementEngine and its IKCCConfig
        /*
        // _movementEngine.ApplyMovement(deltaTime, transform.position + velocity, true);
        */
    }
```

Overall I want to greatly expand on this manual usage of the "mid-level" abstractions of OpenKCC, and I hope that this could be a decent starting point.

**Comment by nicholas-maltbie** [comment]
I think that adding another call to the KCCMovementEngine for stating desired movement with some parameters is probably the best way to handle this situation. 

It already has a reference to the KCCConfig, so it's probably best to pass most of these variables through the config or as properties of the KCCMovementEngine itself. The only parameters that shouldn't be passed would be the movement and delta time (as those could vary depending on how you want to call the function).

I'll add the new api something like this.
Then when making the calls you would simply need to update the apply gravity or apply velocity parameters of the KCC config based on what behavior you would like. The normal default behavior would be to apply gravity whenever the player is either "not grounded" or when the slope of the ground they are walking on is greater than their max walk angle. That way I can remove that redundant attribute `ApplyGravityAttribute` and have it be derived from the grounded state.

```C#
public interface IKCCConfig
{
    public bool ApplyGravity { get; }
    public bool ApplyVelocity { get; }
}

class KCCMovementEngine
{
  public Vector3 Velocity { get; protected set; }
  public IKCCConfig config;

  public void MovePlayer(Vector3 movement, Vector3 deltaTime)
  {
    bool applyGravity = config.ApplyGravity;
    bool allowVelocity = config.AllowVelocity;

    if (ApplyGravity)
    {
      // Update velocity based on gravity
    }

    if (!AllowVelocity)
    {
      Velocity = Vector3.zero;
    }

    // Call other MovePlyaerInternal with this set of parameters...
  }
}
```

**Comment by nicholas-maltbie** [comment]
I wrote up a PR based on this issue (PR #137). Look over the PR and let me know if I left anything out in those APIs you would like added. If the changes look good for now we can merge them and always make more adjustments later.

You can import the code into your project from that branch via git in the unity editor (but not the samples) via the import path.

![image](https://user-images.githubusercontent.com/3240136/213370354-323eb106-3753-4fb6-896c-7131a3aad5f8.png)

```json
{
    "com.nickmaltbie.openkcc": "git+https://github.com/nicholas-maltbie/openkcc.git?path=Packages/com.nickmaltbie.openkcc#nickmaltbie/api-kcc-engine-config-update",
    "com.nickmaltbie.openkcc.netcode": "git+https://github.com/nicholas-maltbie/openkcc.git?path=Packages/com.nickmaltbie.openkcc.netcode#nickmaltbie/api-kcc-engine-config-update",
}
```

**Comment by MTrecozzi** [comment]
**Update:** First Iteration of Usage of new KCCMovementEngine API

**Updated Test Case:** Utilized the New API alongside a Sample Client Script to Implement the Initial Test Case

![Open-KCC-Manual-Usage](https://user-images.githubusercontent.com/47111522/213613476-9f42fbcb-1a8f-4954-9354-4363f128741d.gif)

```c#
void FixedUpdate()
    {
        var deltaTime = Time.fixedDeltaTime;

        var playerInput = playerMovement.ReadValue<Vector2>();

        var desiredVelocity =
            useTestInput ?
                new Vector3(testInput.x, 0, testInput.y) : new Vector3(playerInput.x, 0, playerInput.y);

        desiredVelocity *= RunSpeed;

        Velocity = new Vector3(playerInput.x, Velocity.y, playerInput.y);

        // Direct Control Over Planar Movement
        Velocity.x = desiredVelocity.x;
        Velocity.z = desiredVelocity.z;

        // Apply Jump Velocity if Jump Input is Triggered
        if (_movementEngine.groundedState.OnGround && jumpInput.triggered)
        {
            Velocity.y = JumpForce;
        }

        // Apply Movement to Character based on Calculated Velocity
        ApplyMovement(Velocity, deltaTime);

        // Update Y Velocity Based on Grounded State
        Velocity.y = _movementEngine.groundedState.StandingOnGround ? 0 : Velocity.y + gravity.y * deltaTime;
    }
```

**Comment by MTrecozzi** [comment]
**Update:** First Iteration of original test case using new KCCMovementEngine API

**Updated Test Case:** Utilized the New API alongside a Sample Client Script to Implement the Original Test Case

![Open-KCC-Manual-Usage](https://user-images.githubusercontent.com/47111522/213613476-9f42fbcb-1a8f-4954-9354-4363f128741d.gif)

**Relevant Client Code Snippet:**

```c#
    void FixedUpdate()
    {
        var deltaTime = Time.fixedDeltaTime;

        var playerInput = playerMovement.ReadValue<Vector2>();

        var desiredVelocity =
            useTestInput ?
                new Vector3(testInput.x, 0, testInput.y) : new Vector3(playerInput.x, 0, playerInput.y);

        // Multiply Planar Input by Run Speed
        desiredVelocity *= RunSpeed;

        // Direct Control Over Planar Movement
        Velocity.x = desiredVelocity.x;
        Velocity.z = desiredVelocity.z;

        // Apply Jump Velocity if Jump Input is Triggered
        if (_movementEngine.groundedState.OnGround && jumpInput.triggered)
        {
            Velocity.y = JumpForce;
        }

        // Apply Movement to Character based on Calculated Velocity
        ApplyMovement(Velocity, deltaTime);

        // Update Y Velocity Based on Grounded State
        Velocity.y = _movementEngine.groundedState.StandingOnGround ? 0 : Velocity.y + gravity.y * deltaTime;
    }

    void ApplyMovement(Vector3 velocity, float deltaTime)
    {
        // MovePlayer based on custom velocity
        _movementEngine.MovePlayer(deltaTime, velocity * deltaTime);
    }
```
**Configuration:**
```c#
[System.Serializable]
public class SimpleKCCConfig : IKCCConfig
{

    public bool ApplyGravity => false;

    public bool ApplyVelocity => false;

    public bool SnapPlayerDown => true;
}
```

**Issue 1:** Probably due to my misuse of the API, The Object Phases Through The Ground Before Snapping Back Up When Falling

![Open-KCC-Floor-Falling-Through](https://user-images.githubusercontent.com/47111522/213618411-11c248df-edc4-44c4-8514-be5317fed31a.gif)

**Issue 2:** Jump Force Needs to be Higher than 10 in order for Jump to Actually Occur
- This is probably because I always have "Snap To Ground" returning true, when I'm probably expected to manage the state of that variable on my own

![Open-KCC-Jump](https://user-images.githubusercontent.com/47111522/213618780-e08dbbd8-eea2-4149-a9d1-5964d687f9a7.gif)

**Issue 3:** Related to the above, with a jump force of 10 the character can't jump while moving up stairs, but can jump while moving down stairs

**Notes:**
I wanted to be more detailed in this comment but I ended up having some other work to prioritize. I also would not be surprised if a lot of this friction is just coming from my misusing or misunderstanding the API, so I'm receptive to more insight on that front.

I'll be busy tomorrow, but I hope to resume this experimental usage of the package sometime this weekend. I also have an idea for a potential extension of KCCMovementEngine that would basically allow users to provide it with an "IVelocityProvider", allowing users to define its movement by defining a single method for this functional interface
- It could look something like, 'IVelocityProvider.GetVelocity(ref Vector3 currentVelocity, float deltaTime)'
- My intuition is that by simply asking the user of the KCCMovementEngine for a velocity each frame, the KCCMovementEngine could more tightly handle things such as object snapping and modifying current Velocity while grounded or colliding with something

These are just my scattered thoughts, but I hope to more formally propose something after I have more experience with the package

I'm also expecting that some of my suggestions and intuitions may be off base since I don't fully understand how all the core components work just yet

**Comment by nicholas-maltbie** [comment]
Hey @MTrecozzi , I see what you are saying. I'm going to add PR #137 to the codebase because it has some good changes and bugfixes. However what I think we need to design is a better api for handling how to enable things like `ApplySnap` `ApplyVelocity` and `ApplyGravity` more naturally.

Like you mentioned with the ApplySnap and not having it enabled. We can do a simple check if the player's velocity is upwards (`Proj(Vel, Up) dot Up > 0`) then we know that the player is attempting to move up and should snap them down.

I think I fixed the velocity becoming very small when apply velocity is disabled and just zero out the value.

Moving Up function: https://github.com/nicholas-maltbie/OpenKCC/blob/b970c6a27aa8d96c552afd54c874b7f9a6f96a1b/Packages/com.nickmaltbie.openkcc/OpenKCC/Character/KCCMovementEngine.cs#L78
Zeroing velocity - https://github.com/nicholas-maltbie/OpenKCC/blob/b970c6a27aa8d96c552afd54c874b7f9a6f96a1b/Packages/com.nickmaltbie.openkcc/OpenKCC/Character/KCCMovementEngine.cs#L125

We could keep on adding more features and adjusting until we hit something that works right, but I think it would be better to re-design the API with a new set of requirements in mind.

Let's keep this issue open and add notes about re-design of API after more discussion. 

Would probably be best to keep it to as few classes and interfaces as possible to make it simple to use. So maybe just the KCCMovementEngine and KCCConfig (get rid of the KCCGroundedState to make it simpler and combine it with the KCCConfig).

Then we define all the relationships between the KCCMovementEngine and KCCConfig and allow the user to pass the simple parameters of movement and delta time to the KCCMovementEngine.

**Comment by nicholas-maltbie** [comment]
I wrote up the basic document and made a PR #138 for it. The start of the document basically describes why this is needed. Right now the API wants to satisfy a bunch of different stakeholders and it's doing too many things poorly. For most scenarios you won't want to modify that parameter from it's default. So, let's make the default the right setting in the first place and let the user override it if they want to.

Another thing to consider is constraints on player movement. If the player starts climbing something or holding onto a ledge their movement is now constrained to fewer axes. While if the player is flying in 3D space, their movement has no constraints at all.

here is the list of scenarios: https://github.com/nicholas-maltbie/OpenKCC/blob/nickmaltbie/api-requirements/Documentation/manual/kcc-design/kcc-user-stories.md#scenarios
and the design proposal: https://github.com/nicholas-maltbie/OpenKCC/blob/nickmaltbie/api-requirements/Documentation/manual/kcc-design/kcc-user-stories.md#designing-an-api

If you or anyone else has any suggestions, feel free to leave them in this issue.

**Comment by nicholas-maltbie** [comment]
Opened PR #142 to address this issue :)

**Comment by nicholas-maltbie** [comment]
I think I feel content with the new version of the api.

Closed PR #142 to overhaul the api and significantly improved documentation, testing, and exposed api parameters to make the newer version more accessible. 

I opened a PR to update the project version to 1.3.0 and will create a release once we have PR #150 merged!

If you have any other suggestions or comments let me know. We can always follow up in a future patch or release as well.
